### PR TITLE
Add product to cart

### PR DESCRIPTION
**Pull request by trinhdvt**
## Related Tickets
- [#46909](https://edu-redmine.sun-asterisk.vn/issues/46909)

## WHAT (optional)
- Add js-library `toastr`
- Implement Add Product to Cart feature

## HOW
- At Product's detail page, user can select quantity then click on `Add to Cart` button to add product. An `ajax` request will be sent to controller
- Cart is stored in cookies will expire time is 90 days
- If Product is already existed in Cart then update old quantity else append it into.
- If request success, display a toast message and update cart's amount in UI

## Evidence (Screenshot or Video)

Ảnh chụp Màn hình 2022-03-16 lúc 08 40 45
Ảnh chụp Màn hình 2022-03-16 lúc 08 45 42

### CODE REVIEW

**Review by lcko1012 — APPROVED** [review]
ok

**Comment by trinhdvt** [comment]
ready

File: app/controllers/carts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  private
+
+  def get_quantity
+    @quantity = params.require(:cart)
+                      .values_at(:quantity)
+                      .first
+                      .to_i
+    solve_invalid_req t(".invalid_qtt") if @quantity.zero?
```
Comment (vanvtt-0952) [inline app/controllers/carts_controller.rb]:
```suggestion
    @quantity = params.dig(:cart, :quantity) || 1
```

File: app/controllers/carts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  def get_product
+    product_id = params.require(:cart)
+                       .values_at(:product_id)
+                       .first
```
Comment (vanvtt-0952) [inline app/controllers/carts_controller.rb]: Chị cứ thắc mắc sao cần cái `.first` nhỉ

File: app/controllers/carts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    @product = Product.find_by id: product_id
+    solve_invalid_req t("products.not_found") unless @product
```
Comment (vanvtt-0952) [inline app/controllers/carts_controller.rb]:
```suggestion
    return if  @product
    
    solve_invalid_req t("products.not_found")
```

File: app/controllers/carts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+class CartsController < ApplicationController
+  before_action :get_product, :get_quantity, only: :create
```
Comment (vanvtt-0952) [inline app/controllers/carts_controller.rb]:
```suggestion
  before_action :handle_param_quantity, :load_product, only: :create
```

File: app/controllers/carts_controller.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    end
+
+    respond_to do |format|
+      format.js
+    end
+  end
+
+  private
+
+  def get_quantity
```
Comment (vanvtt-0952) [inline app/controllers/carts_controller.rb]:
```suggestion
  def handle_param_quantity
```

File: app/controllers/carts_controller.rb (hunk 2 above)
Comment (trinhdvt) [reply app/controllers/carts_controller.rb]: bị dư c ạ 🥲  e quên sửa

**Pull request merged**